File: python/tests/test_caching.py
import time

import pytest

from output import Airtable


@pytest.fixture(scope="module")
def airtable_cached():
    return Airtable(cache_seconds=60)


@pytest.fixture(scope="module")
def airtable_uncached():
    return Airtable()


def new_primary_dict(fields=None):
    return {"fields": fields or {}}


def new_secondary_dict(fields=None):
    return {"fields": fields or {}}


class TestCacheHitReferenceEquality:
    id: str

    def test_create(self, airtable_cached):
        record = new_primary_dict({"Primary Key": "Cache Hit Test"})
        created = airtable_cached.primary.dict.create(record)
        self.__class__.id = created["id"]
        assert created["id"]

    def test_cache_hit_dict(self, airtable_cached):
        first = airtable_cached.primary.dict.get(self.id)
        second = airtable_cached.primary.dict.get(self.id)
        assert first is second

    def test_cache_hit_orm(self, airtable_cached):
        first = airtable_cached.primary.get(self.id)
        second = airtable_cached.primary.get(self.id)
        assert first is second

    def test_cleanup(self, airtable_cached):
        airtable_cached.primary.dict.delete(record_id=self.id)


class TestCacheDisabledByDefault:
    def test_cache_seconds_is_zero(self, airtable_uncached):
        assert airtable_uncached._cache_seconds == 0


class TestMutationInvalidation:
    id: str
    id2: str

    def test_create_record(self, airtable_cached):
        record = new_primary_dict({"Primary Key": "Mutation Test"})
        created = airtable_cached.primary.dict.create(record)
        self.__class__.id = created["id"]

    def test_get_populates_cache(self, airtable_cached):
        airtable_cached.primary.dict.get(self.id)
        assert len(airtable_cached.primary.dict._cache) > 0

    def test_create_invalidates_cache(self, airtable_cached):
        record2 = new_primary_dict({"Primary Key": "Mutation Test 2"})
        created2 = airtable_cached.primary.dict.create(record2)
        self.__class__.id2 = created2["id"]
        assert len(airtable_cached.primary.dict._cache) == 0

    def test_update_invalidates_cache(self, airtable_cached):
        airtable_cached.primary.dict.get(self.id)
        assert len(airtable_cached.primary.dict._cache) > 0
        r = airtable_cached.primary.dict.get(self.id)
        r["fields"]["Primary Key"] = "Mutation Updated"
        airtable_cached.primary.dict.update(r)
        assert len(airtable_cached.primary.dict._cache) == 0

    def test_delete_invalidates_cache(self, airtable_cached):
        airtable_cached.primary.dict.get(self.id)
        assert len(airtable_cached.primary.dict._cache) > 0
        airtable_cached.primary.dict.delete(record_id=self.id2)
        assert len(airtable_cached.primary.dict._cache) == 0

    def test_cleanup(self, airtable_cached):
        airtable_cached.primary.dict.delete(record_id=self.id)


class TestManualInvalidation:
    id: str

    def test_create(self, airtable_cached):
        record = new_primary_dict({"Primary Key": "Manual Invalidation Test"})
        created = airtable_cached.primary.dict.create(record)
        self.__class__.id = created["id"]

    def test_populate_cache(self, airtable_cached):
        airtable_cached.primary.dict.get(self.id)
        assert len(airtable_cached.primary.dict._cache) > 0

    def test_invalidate_clears_cache(self, airtable_cached):
        airtable_cached.primary.invalidate_cache()
        assert len(airtable_cached.primary._cache) == 0
        assert len(airtable_cached.primary.dict._cache) == 0

    def test_get_still_works(self, airtable_cached):
        result = airtable_cached.primary.dict.get(self.id)
        assert result["id"] == self.id

    def test_cleanup(self, airtable_cached):
        airtable_cached.primary.dict.delete(record_id=self.id)


class TestMainClientCascadeInvalidation:
    primary_id: str
    secondary_id: str

    def test_create_records(self, airtable_cached):
        p = new_primary_dict({"Primary Key": "Cascade Primary"})
        created_p = airtable_cached.primary.dict.create(p)
        self.__class__.primary_id = created_p["id"]

        s = new_secondary_dict({"Name": "Cascade Secondary"})
        created_s = airtable_cached.secondary.dict.create(s)
        self.__class__.secondary_id = created_s["id"]

    def test_populate_caches(self, airtable_cached):
        airtable_cached.primary.dict.get(self.primary_id)
        airtable_cached.secondary.dict.get(self.secondary_id)
        assert len(airtable_cached.primary.dict._cache) > 0
        assert len(airtable_cached.secondary.dict._cache) > 0

    def test_cascade_invalidation(self, airtable_cached):
        airtable_cached.invalidate_cache()
        assert len(airtable_cached.primary._cache) == 0
        assert len(airtable_cached.primary.dict._cache) == 0
        assert len(airtable_cached.secondary._cache) == 0
        assert len(airtable_cached.secondary.dict._cache) == 0

    def test_cleanup(self, airtable_cached):
        airtable_cached.primary.dict.delete(record_id=self.primary_id)
        airtable_cached.secondary.dict.delete(record_id=self.secondary_id)


class TestCacheExpiry:
    id: str
    original_cache_seconds: int

    def test_create(self, airtable_cached):
        record = new_primary_dict({"Primary Key": "Expiry Test"})
        created = airtable_cached.primary.dict.create(record)
        self.__class__.id = created["id"]

    def test_cache_then_expire(self, airtable_cached):
        # Temporarily set a short TTL
        self.__class__.original_cache_seconds = airtable_cached.primary.dict._cache_seconds
        airtable_cached.primary.dict._cache_seconds = 2

        first = airtable_cached.primary.dict.get(self.id)
        cached = airtable_cached.primary.dict.get(self.id)
        assert first is cached

        time.sleep(3)

        after_expiry = airtable_cached.primary.dict.get(self.id)
        assert after_expiry is not first

        # Restore original TTL
        airtable_cached.primary.dict._cache_seconds = self.original_cache_seconds

    def test_cleanup(self, airtable_cached):
        airtable_cached.primary.dict.delete(record_id=self.id)


class TestCacheServesStaleData:
    id: str

    def test_create(self, airtable_cached):
        record = new_primary_dict({"Primary Key": "Stale Data Test"})
        created = airtable_cached.primary.dict.create(record)
        self.__class__.id = created["id"]

    def test_stale_data(self, airtable_cached):
        # Populate cache with original value
        first = airtable_cached.primary.dict.get(self.id)
        assert first["fields"]["Primary Key"] == "Stale Data Test"

        # Update via underlying pyairtable API directly (bypasses cache invalidation)
        airtable_cached.primary._table.update(self.id, {"Primary Key": "Updated Directly"})

        # Cached instance should return stale data
        stale = airtable_cached.primary.dict.get(self.id)
        assert stale["fields"]["Primary Key"] == "Stale Data Test"

    def test_cleanup(self, airtable_cached):
        airtable_cached.primary.dict.delete(record_id=self.id)


class TestDifferentParamsDifferentCacheKeys:
    id1: str
    id2: str

    def test_create(self, airtable_cached):
        r1 = new_primary_dict({"Primary Key": "Cache Key 1"})
        r2 = new_primary_dict({"Primary Key": "Cache Key 2"})
        c1 = airtable_cached.primary.dict.create(r1)
        c2 = airtable_cached.primary.dict.create(r2)
        self.__class__.id1 = c1["id"]
        self.__class__.id2 = c2["id"]

    def test_different_keys(self, airtable_cached):
        airtable_cached.primary.invalidate_cache()
        get1 = airtable_cached.primary.dict.get(self.id1)
        get2 = airtable_cached.primary.dict.get(self.id2)
        assert len(airtable_cached.primary.dict._cache) == 2

        re_get1 = airtable_cached.primary.dict.get(self.id1)
        re_get2 = airtable_cached.primary.dict.get(self.id2)
        assert re_get1 is get1
        assert re_get2 is get2

    def test_cleanup(self, airtable_cached):
        airtable_cached.primary.dict.delete(record_id=self.id1)
        airtable_cached.primary.dict.delete(record_id=self.id2)


class TestBatchGetCaching:
    id1: str
    id2: str

    def test_create(self, airtable_cached):
        r1 = new_primary_dict({"Primary Key": "Batch Cache 1"})
        r2 = new_primary_dict({"Primary Key": "Batch Cache 2"})
        c1 = airtable_cached.primary.dict.create(r1)
        c2 = airtable_cached.primary.dict.create(r2)
        self.__class__.id1 = c1["id"]
        self.__class__.id2 = c2["id"]

    def test_batch_cached_separately_from_single(self, airtable_cached):
        airtable_cached.primary.invalidate_cache()

        # Batch get caches as one entry
        batch = airtable_cached.primary.dict.get([self.id1, self.id2])
        assert len(airtable_cached.primary.dict._cache) == 1

        # Single get creates a separate cache entry
        single = airtable_cached.primary.dict.get(self.id1)
        assert len(airtable_cached.primary.dict._cache) == 2

        # Both return cached results on re-get
        batch2 = airtable_cached.primary.dict.get([self.id1, self.id2])
        single2 = airtable_cached.primary.dict.get(self.id1)
        assert batch2 is batch
        assert single2 is single

    def test_cleanup(self, airtable_cached):
        airtable_cached.primary.dict.delete(record_id=self.id1)
        airtable_cached.primary.dict.delete(record_id=self.id2)


class TestFieldsOptionAffectsCacheKey:
    id: str

    def test_create(self, airtable_cached):
        record = new_primary_dict({"Primary Key": "Fields Cache Test", "Single Line Text": "Hello"})
        created = airtable_cached.primary.dict.create(record)
        self.__class__.id = created["id"]

    def test_different_fields_different_cache_entries(self, airtable_cached):
        airtable_cached.primary.invalidate_cache()

        # Get with no fields filter
        full = airtable_cached.primary.dict.get(self.id)
        assert len(airtable_cached.primary.dict._cache) == 1

        # Get with fields filter creates a separate cache entry
        partial = airtable_cached.primary.dict.get(self.id, fields=["Primary Key"])
        assert len(airtable_cached.primary.dict._cache) == 2

        # Both return cached results on re-get
        full2 = airtable_cached.primary.dict.get(self.id)
        partial2 = airtable_cached.primary.dict.get(self.id, fields=["Primary Key"])
        assert full2 is full
        assert partial2 is partial

        # Partial result should not have the extra field
        assert "Single Line Text" not in partial["fields"]

    def test_cleanup(self, airtable_cached):
        airtable_cached.primary.dict.delete(record_id=self.id)


class TestFormulaQueryCaching:
    id: str

    def test_create(self, airtable_cached):
        record = new_primary_dict({"Primary Key": "Formula Cache Test"})
        created = airtable_cached.primary.dict.create(record)
        self.__class__.id = created["id"]

    def test_formula_query_is_cached(self, airtable_cached):
        airtable_cached.primary.invalidate_cache()

        formula = "{Primary Key} = 'Formula Cache Test'"
        first = airtable_cached.primary.dict.get(formula=formula)
        assert len(airtable_cached.primary.dict._cache) == 1

        second = airtable_cached.primary.dict.get(formula=formula)
        assert first is second

    def test_different_formulas_different_cache_entries(self, airtable_cached):
        airtable_cached.primary.invalidate_cache()

        f1 = "{Primary Key} = 'Formula Cache Test'"
        f2 = "{Primary Key} != 'nonexistent'"
        airtable_cached.primary.dict.get(formula=f1)
        airtable_cached.primary.dict.get(formula=f2)
        assert len(airtable_cached.primary.dict._cache) == 2

    def test_cleanup(self, airtable_cached):
        airtable_cached.primary.dict.delete(record_id=self.id)


class TestOrmAndDictIndependentCaches:
    id: str

    def test_create(self, airtable_cached):
        record = new_primary_dict({"Primary Key": "Independent Cache Test"})
        created = airtable_cached.primary.dict.create(record)
        self.__class__.id = created["id"]

    def test_orm_cache_independent_of_dict(self, airtable_cached):
        airtable_cached.primary.invalidate_cache()

        # ORM get populates ORM cache only
        airtable_cached.primary.get(self.id)
        assert len(airtable_cached.primary._cache) > 0
        assert len(airtable_cached.primary.dict._cache) == 0

        # Dict get populates dict cache only
        airtable_cached.primary.dict.get(self.id)
        assert len(airtable_cached.primary._cache) > 0
        assert len(airtable_cached.primary.dict._cache) > 0

    def test_cleanup(self, airtable_cached):
        airtable_cached.primary.dict.delete(record_id=self.id)

File: rust/tests/test_caching.rs
use myairtable_tests::*;

fn setup_cached() -> Airtable {
    dotenvy::from_filename(".env").ok();
    let api_key = std::env::var("AIRTABLE_API_KEY").expect("AIRTABLE_API_KEY must be set");
    let base_id = std::env::var("AIRTABLE_BASE_ID").expect("AIRTABLE_BASE_ID must be set");
    Airtable::with_cache(&api_key, &base_id, 60)
}

fn setup_uncached() -> Airtable {
    dotenvy::from_filename(".env").ok();
    let api_key = std::env::var("AIRTABLE_API_KEY").expect("AIRTABLE_API_KEY must be set");
    let base_id = std::env::var("AIRTABLE_BASE_ID").expect("AIRTABLE_BASE_ID must be set");
    Airtable::new(&api_key, &base_id)
}

// =============================================================================
// Cache hit — second get returns same data without API call
// =============================================================================

#[tokio::test]
async fn cache_hit_returns_same_data() {
    let at = setup_cached();

    let fields = CreatePrimaryModel {
        primary_key: Some("Cache Hit Test".to_string()),
        ..Default::default()
    };
    let created = at.primary.create_one(&fields).await.unwrap();
    let id = created.id.as_deref().unwrap().to_string();

    // First get populates cache
    let first = at.primary.get_one(&id).await.unwrap();
    // Second get should return cached data
    let second = at.primary.get_one(&id).await.unwrap();

    assert_eq!(first.primary_key, second.primary_key);
    assert_eq!(first.id, second.id);

    at.primary.delete_one(&id).await.unwrap();
}

// =============================================================================
// Cache disabled by default
// =============================================================================

#[tokio::test]
async fn cache_disabled_by_default() {
    let at = setup_uncached();

    let fields = CreatePrimaryModel {
        primary_key: Some("No Cache Test".to_string()),
        ..Default::default()
    };
    let created = at.primary.create_one(&fields).await.unwrap();
    let id = created.id.as_deref().unwrap().to_string();

    // Get should work but not cache (cache_seconds = 0)
    let result = at.primary.get_one(&id).await.unwrap();
    assert_eq!(result.primary_key.as_deref(), Some("No Cache Test"));

    at.primary.delete_one(&id).await.unwrap();
}

// =============================================================================
// Mutation invalidation
// =============================================================================

#[tokio::test]
async fn create_invalidates_cache() {
    let at = setup_cached();

    let fields = CreatePrimaryModel {
        primary_key: Some("Mutation Test 1".to_string()),
        ..Default::default()
    };
    let created = at.primary.create_one(&fields).await.unwrap();
    let id = created.id.as_deref().unwrap().to_string();

    // Populate cache
    at.primary.get_one(&id).await.unwrap();

    // Create should invalidate cache
    let fields2 = CreatePrimaryModel {
        primary_key: Some("Mutation Test 2".to_string()),
        ..Default::default()
    };
    let created2 = at.primary.create_one(&fields2).await.unwrap();
    let id2 = created2.id.as_deref().unwrap().to_string();

    // After invalidation, get should hit API (not cache)
    let result = at.primary.get_one(&id).await.unwrap();
    assert_eq!(result.primary_key.as_deref(), Some("Mutation Test 1"));

    at.primary.delete_one(&id).await.unwrap();
    at.primary.delete_one(&id2).await.unwrap();
}

#[tokio::test]
async fn update_invalidates_cache() {
    let at = setup_cached();

    let fields = CreatePrimaryModel {
        primary_key: Some("Update Invalidation".to_string()),
        ..Default::default()
    };
    let created = at.primary.create_one(&fields).await.unwrap();
    let id = created.id.as_deref().unwrap().to_string();

    // Populate cache
    at.primary.get_one(&id).await.unwrap();

    // Update invalidates cache
    let update = CreatePrimaryModel {
        primary_key: Some("Updated Value".to_string()),
        ..Default::default()
    };
    at.primary.update_one(&id, &update).await.unwrap();

    // Next get should see updated value (cache was invalidated)
    let result = at.primary.get_one(&id).await.unwrap();
    assert_eq!(result.primary_key.as_deref(), Some("Updated Value"));

    at.primary.delete_one(&id).await.unwrap();
}

#[tokio::test]
async fn delete_invalidates_cache() {
    let at = setup_cached();

    let fields = CreatePrimaryModel {
        primary_key: Some("Delete Invalidation".to_string()),
        ..Default::default()
    };
    let created = at.primary.create_one(&fields).await.unwrap();
    let id = created.id.as_deref().unwrap().to_string();

    // Populate cache
    at.primary.get_one(&id).await.unwrap();

    // Delete invalidates cache
    at.primary.delete_one(&id).await.unwrap();

    // Next get should fail (record deleted, cache cleared)
    assert!(at.primary.get_one(&id).await.is_err());
}

// =============================================================================
// Manual invalidation
// =============================================================================

#[tokio::test]
async fn manual_invalidate_cache() {
    let at = setup_cached();

    let fields = CreatePrimaryModel {
        primary_key: Some("Manual Invalidation".to_string()),
        ..Default::default()
    };
    let created = at.primary.create_one(&fields).await.unwrap();
    let id = created.id.as_deref().unwrap().to_string();

    // Populate cache
    at.primary.get_one(&id).await.unwrap();

    // Manual invalidation
    at.primary.invalidate_cache();

    // Get still works (re-fetches from API)
    let result = at.primary.get_one(&id).await.unwrap();
    assert_eq!(result.id.as_deref(), Some(id.as_str()));

    at.primary.delete_one(&id).await.unwrap();
}

// =============================================================================
// Cache expiry
// =============================================================================

#[tokio::test]
async fn cache_expires_after_ttl() {
    let mut at = setup_uncached();
    at.primary.set_cache_seconds(1); // 1 second TTL

    let fields = CreatePrimaryModel {
        primary_key: Some("Expiry Test".to_string()),
        ..Default::default()
    };
    let created = at.primary.create_one(&fields).await.unwrap();
    let id = created.id.as_deref().unwrap().to_string();

    // Populate cache
    let first = at.primary.get_one(&id).await.unwrap();
    assert_eq!(first.primary_key.as_deref(), Some("Expiry Test"));

    // Wait for TTL to expire
    tokio::time::sleep(std::time::Duration::from_secs(2)).await;

    // After expiry, should re-fetch from API
    let after = at.primary.get_one(&id).await.unwrap();
    assert_eq!(after.primary_key.as_deref(), Some("Expiry Test"));

    at.primary.delete_one(&id).await.unwrap();
}

// =============================================================================
// Cache serves stale data
// =============================================================================

#[tokio::test]
async fn cache_serves_stale_data() {
    let at = setup_cached();

    let fields = CreatePrimaryModel {
        primary_key: Some("Stale Data Test".to_string()),
        ..Default::default()
    };
    let created = at.primary.create_one(&fields).await.unwrap();
    let id = created.id.as_deref().unwrap().to_string();

    // Populate cache
    let first = at.primary.get_one(&id).await.unwrap();
    assert_eq!(first.primary_key.as_deref(), Some("Stale Data Test"));

    // Update via dict layer (bypasses ORM cache invalidation)
    let mut update_fields = Fields::new();
    update_fields.set(PrimaryFields::PRIMARY_KEY_ID, "Updated Directly");
    at.primary
        .dict
        .update_one(&id, &update_fields, true)
        .await
        .unwrap();

    // ORM cache should still return stale data
    let stale = at.primary.get_one(&id).await.unwrap();
    assert_eq!(stale.primary_key.as_deref(), Some("Stale Data Test"));

    at.primary.delete_one(&id).await.unwrap();
}

// =============================================================================
// Different parameters produce different cache keys
// =============================================================================

#[tokio::test]
async fn different_queries_cached_independently() {
    let at = setup_cached();

    let fields1 = CreatePrimaryModel {
        primary_key: Some("Cache Key 1".to_string()),
        ..Default::default()
    };
    let fields2 = CreatePrimaryModel {
        primary_key: Some("Cache Key 2".to_string()),
        ..Default::default()
    };
    let c1 = at.primary.create_one(&fields1).await.unwrap();
    let c2 = at.primary.create_one(&fields2).await.unwrap();
    let id1 = c1.id.as_deref().unwrap().to_string();
    let id2 = c2.id.as_deref().unwrap().to_string();

    // Each get populates a separate cache entry
    let get1 = at.primary.get_one(&id1).await.unwrap();
    let get2 = at.primary.get_one(&id2).await.unwrap();

    // Re-get returns correct cached data (not mixed up)
    let reget1 = at.primary.get_one(&id1).await.unwrap();
    let reget2 = at.primary.get_one(&id2).await.unwrap();
    assert_eq!(reget1.primary_key, get1.primary_key);
    assert_eq!(reget2.primary_key, get2.primary_key);

    at.primary.delete_one(&id1).await.unwrap();
    at.primary.delete_one(&id2).await.unwrap();
}

// =============================================================================
// Formula query caching
// =============================================================================

#[tokio::test]
async fn formula_query_cached() {
    let at = setup_cached();

    let fields = CreatePrimaryModel {
        primary_key: Some("Formula Cache Test".to_string()),
        ..Default::default()
    };
    let created = at.primary.create_one(&fields).await.unwrap();
    let id = created.id.as_deref().unwrap().to_string();

    // Query with formula
    let query = AirtableQuery::new().formula(PrimaryModel::F.primary_key.equals(
        "Formula Cache Test",
        true,
        false,
    ));
    let first = at.primary.get_many(&query).await.unwrap();
    assert!(!first.is_empty());

    // Second query with same formula returns cached data
    let second = at.primary.get_many(&query).await.unwrap();
    assert_eq!(first.len(), second.len());

    at.primary.delete_one(&id).await.unwrap();
}
